### PR TITLE
Ingest of VHS 1256b data

### LINKED ISSUE #284: VHS 1256 b spectral data ingest
Ingesting VHS 1256 b spectral data...

JHK band data from Gauza+2015, L band data from Miles+2018.

[vhs1256b_spectra_Figure8_Gauza2015.txt](https://github.com/SIMPLE-AstroDB/SIMPLE-db/files/9455987/vhs1256b_spectra_Figure8_Gauza2015.txt)
[vhs1256b_spectra_Figure8_Miles2018.txt](https://github.com/SIMPLE-AstroDB/SIMPLE-db/files/9455988/vhs1256b_spectra_Figure8_Miles2018.txt)

### PR DESCRIPTION
**Pull request by kelle**
will close #284.

### CODE REVIEW

**Comment by kelle** [comment]
I haven't finished converting the spectra, but here's an ingest script for all the other data associated with VHS1256B.  It also includes improvements to various ingest scripts. Please review! @dr-rodriguez @SherelynA @NiallWhiteford .

**Review by NiallWhiteford — COMMENTED** [review]
Went through ingest. Looks fine to me!

File: scripts/ingests/ingest_utils.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from astroquery.simbad import Simbad
 from astropy.coordinates import SkyCoord
 import astropy.units as u
+from astroquery.gaia import Gaia
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:8]:
```suggestion
from typing import List, Union, Optional
```

File: scripts/ingests/ingest_utils.py [hunk 2]
```diff
@@ -512,34 +513,44 @@ def convert_spt_string_to_code(spectral_types):
...
+    else:
+        n_sources = len(sources)
 
     # Convert single element input value to list
     if isinstance(plx_refs, str):
         plx_refs = [plx_refs] * n_sources
 
+    if isinstance(comments, str):
+        comments = [comments] * n_sources
+    elif isinstance(comments, type(None)):
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:551]:
```suggestion
    elif comments is None:
```

File: scripts/ingests/ingest_utils.py [hunk 3]
```diff
@@ -792,14 +805,21 @@ def ingest_photometry(db, sources, bands, magnitudes, magnitude_errors, referenc
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:805]: Anything that could possibly be `None`, I would do `Optional[Union[List[str], str]]` else `Union[List[str], str]]`

File: scripts/ingests/ingest_utils.py [hunk 4]
```diff
@@ -1214,3 +1234,76 @@ def ingest_instrument(db, telescope=None, instrument=None, mode=None):
                 raise SimpleError(msg + '\n' + str(e))
 
     return
+
+
+def get_gaiadr3(gaia_id, verbose=True):
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:1239]:
```suggestion
def get_gaiadr3(gaia_id: int, verbose=True):
```

File: scripts/ingests/ingest_utils.py [hunk 5]
```diff
@@ -1214,3 +1234,76 @@ def ingest_instrument(db, telescope=None, instrument=None, mode=None):
...
+    # Currently setup just to query one source
+    # TODO: add some debug and info messages
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:1241]:
```suggestion
    # TODO: add some debug and info messages
    if not isinstance(gaia_id, int):
        raise ValueError('Gaia ID must be an integer')
```

File: scripts/ingests/ingest_utils.py [hunk 6]
```diff
@@ -1214,3 +1234,76 @@ def ingest_instrument(db, telescope=None, instrument=None, mode=None):
...
+    gaia_query_string = f"SELECT " \
+                        f"parallax, parallax_error, " \
+                        f"pmra, pmra_error, pmdec, pmdec_error " \
+                        f"phot_g_mean_flux, phot_g_mean_flux_error, phot_g_mean_mag," \
+                        f"phot_rp_mean_flux, phot_rp_mean_flux_error, phot_rp_mean_mag" \
+                        f"FROM gaiadr3.gaia_source WHERE " \
+                        f"gaiadr3.gaia_source.source_id = '{gaia_id}'"
+    job_gaia_query = Gaia.launch_job(gaia_query_string, verbose=verbose)
+
+    gaia_data = job_gaia_query.get_results()
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:1251]: Add a handle here for if something breaks in the launch job or gets results, either raise a SimpleError or pass and leave a debug message, depending on intended use case

File: scripts/ingests/ingest_utils.py [hunk 7]
```diff
@@ -1214,3 +1234,76 @@ def ingest_instrument(db, telescope=None, instrument=None, mode=None):
...
+        -2.5 / np.log(10) * gaia_rp_phot['phot_rp_mean_flux_error'] / gaia_rp_phot['phot_rp_mean_flux'])
+
+    if ref == 'GaiaDR2':
+        g_band_name = 'GAIA2.G'
+        rp_band_name = 'GAIA2.Grp'
+    elif ref == 'GaiaEDR3' or ref == 'GaiaDR3':
+        g_band_name = 'GAIA3.G'
+        rp_band_name = 'GAIA3.Grp'
+    else:
+        raise Exception
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:1279]:
```suggestion
        raise SimpleError("your message here")
```

File: scripts/ingests/ingest_utils.py [hunk 8]
```diff
@@ -1214,3 +1234,76 @@ def ingest_instrument(db, telescope=None, instrument=None, mode=None):
...
+    ingest_photometry(db, sources, g_band_name, gaia_g_phot['phot_g_mean_mag'], gaia_g_phot['g_unc'],
+                      ref, ucds='em.opt', telescope='Gaia', instrument='Gaia')
+
+    ingest_photometry(db, sources, rp_band_name, gaia_rp_phot['phot_rp_mean_mag'],
+                      gaia_rp_phot['rp_unc'], ref, ucds='em.opt.R', telescope='Gaia', instrument='Gaia')
+
+    return
+
+
+def ingest_gaia_parallaxes(db, sources, gaia_data, ref):
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:1290]: type hinting

File: scripts/ingests/ingest_utils.py [hunk 9]
```diff
@@ -1214,3 +1234,76 @@ def ingest_instrument(db, telescope=None, instrument=None, mode=None):
...
+    # TODO write some tests
+    unmasked_pi = np.logical_not(gaia_data['parallax'].mask).nonzero()
+    gaia_parallaxes = gaia_data[unmasked_pi]['parallax', 'parallax_error']
+
+    ingest_parallaxes(db, sources, gaia_parallaxes['parallax'],
+                      gaia_parallaxes['parallax_error'], ref)
+
+
+def ingest_gaia_pms(db, sources, gaia_data, ref):
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_utils.py:1299]: type hinting

File: scripts/ingests/ingest_vhs1256b.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+vista_magnitudes = [18.558, 17.136, 15.777, 14.665]
+vista_mag_unc = [0.051, 0.020, 0.015, 0.010]
+vista_epoch = Time('2011-07-01')
+vista_epoch.format = 'decimalyear'
+
+ingest_photometry(db, sources, vista_bands, vista_magnitudes, vista_mag_unc, 'Gauz15',
+                  telescope='VISTA', instrument='VIRCAM', epoch=vista_epoch.value.astype(np.float32))
+
+# Ingest Gaia parallax from primary 2MASS J12560215-1257217, Gaia DR3 3526198184723289472
+gaia_data = get_gaiadr3('3526198184723289472', verbose=False)
```
Comment (Will-Cooper) [inline scripts/ingests/ingest_vhs1256b.py:164]: I don't like passing gaia IDs as strings, generally

**Review by Will-Cooper — COMMENTED** [review]
Some minor bits complaints but otherwise looks good

**Comment by kelle** [comment]
Put a bunch of Will's comments in #295

**Review by Will-Cooper — APPROVED** [review]
see #295 for further changes

**Pull request merged**